File: spring-backend/src/main/java/com/simmortal/notification/NotificationController.java
package com.simmortal.notification;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("notification")
public class NotificationController {
  private final NotificationService notificationService;

  public NotificationController(NotificationService notificationService) {
    this.notificationService = notificationService;
  }

  @GetMapping
  public Object getNotifications(
      @RequestHeader("X-User-Id") String userId,
      @RequestParam(value = "cursor", required = false) String cursor
  ) {
    return notificationService.getPaginatedNotifications(userId, cursor);
  }

  @PatchMapping("{notificationId}/read")
  public Object markNotificationRead(
      @RequestHeader("X-User-Id") String userId,
      @PathVariable String notificationId
  ) {
    return notificationService.markNotificationRead(userId, notificationId);
  }
}

File: Backend-main/src/notification/notification.controller.ts
import { Controller, Get, Param, Patch, Query } from '@nestjs/common';
import { Session } from 'supertokens-nestjs';
import { SessionContainer } from 'supertokens-node/recipe/session';
import { NotificationService } from './notification.service';

@Controller('notification')
export class NotificationController {
  constructor(private readonly notificationService: NotificationService) {}

  @Get()
  async getNotifications(@Session() session: SessionContainer, @Query('cursor') cursor?: string) {
    const userId = session.getUserId();
    return this.notificationService.getPaginatedNotifications(userId, cursor);
  }

  @Patch(':notificationId/read')
  async markNotificationRead(
    @Session() session: SessionContainer,
    @Param('notificationId') notificationId: string,
  ) {
    const userId = session.getUserId();
    return this.notificationService.markNotificationRead(userId, notificationId);
  }
}
